src/RequestHandler/inprocess/inprocessapplication.cpp

```diff
@@ -57,49 +56,54 @@ IN_PROCESS_APPLICATION::ShutDown(
     VOID
 )
 {
-    HANDLE   hThread = NULL;
-    HRESULT  hr = S_OK;
-    DWORD    dwThreadStatus = 0;
-    DWORD    dwTimeout = m_pConfig->QueryShutdownTimeLimitInMS();
+    HRESULT hr = S_OK;
+    try
+    {    
+        CHandle  hThread;
+        DWORD    dwThreadStatus = 0;
+        DWORD    dwTimeout = m_pConfig->QueryShutdownTimeLimitInMS();
 
-    if (IsDebuggerPresent())
-    {
-        dwTimeout = INFINITE;
-    }
+        if (IsDebuggerPresent())
+        {
+            dwTimeout = INFINITE;
+        }
 
-    hThread = CreateThread(
-        NULL,       // default security attributes
-        0,          // default stack size
-        (LPTHREAD_START_ROUTINE)DoShutDown,
-        this,       // thread function arguments
-        0,          // default creation flags
-        NULL);      // receive thread identifier
+        hThread.Attach(CreateThread(
+            NULL,       // default security attributes
+            0,          // default stack size
+            (LPTHREAD_START_ROUTINE)DoShutDown,
+            this,       // thread function arguments
+            0,          // default creation flags
+            NULL));      // receive thread identifier
 
-    if (hThread == NULL)
-    {
-        hr = HRESULT_FROM_WIN32(GetLastError());
-        goto Finished;
-    }
+        if ((HANDLE)hThread == NULL)
+        {
+            hr = HRESULT_FROM_WIN32(GetLastError());
+            goto Finished;
+        }
 
-    if (WaitForSingleObject(hThread, dwTimeout) != WAIT_OBJECT_0)
-    {
-        // if the thread is still running, we need kill it first before exit to avoid AV
-        if (GetExitCodeThread(m_hThread, &dwThreadStatus) != 0 && dwThreadStatus == STILL_ACTIVE)
+        if (WaitForSingleObject(hThread, dwTimeout) != WAIT_OBJECT_0)
         {
-            // Calling back into managed at this point is prone to have AVs
-            // Calling terminate thread here may be our best solution.
-            TerminateThread(hThread, STATUS_CONTROL_C_EXIT);
-            hr = HRESULT_FROM_WIN32(ERROR_TIMEOUT);
+            // if the thread is still running, we need kill it first before exit to avoid AV
+            if (GetExitCodeThread(m_hThread, &dwThreadStatus) != 0 && dwThreadStatus == STILL_ACTIVE)
+            {
+                // Calling back into managed at this point is prone to have AVs
+                // Calling terminate thread here may be our best solution.
+                TerminateThread(hThread, STATUS_CONTROL_C_EXIT);
+                hr = HRESULT_FROM_WIN32(ERROR_TIMEOUT);
+            }
         }
     }
-
-Finished:
-
-    if (hThread != NULL)
+    catch (const std::system_error& ex)
     {
-        CloseHandle(hThread);
+        hr = ex.code().value();
+    }
+    catch (...)
+    {
+        hr = ERROR_EXCEPTION_IN_SERVICE;
```

Comment:
Anything that starts with ERROR_* needs to be wrapped HRESULT_FROM_WIN32(). One good macro we have XD